odl/phantom/geometric.py

```diff
@@ -703,6 +704,232 @@ def ellipsoid_phantom(space, ellipsoids, min_pt=None, max_pt=None):
+        x0 = cuboid[3]
+        y0 = cuboid[4]
+        theta = cuboid[5]
+
+        scales = [1 / a_squared, 1 / b_squared]
+        center = (np.array([x0, y0]) + 1.0) / 2.0
+
+        # Create the offset x,y and z values for the grid
+        if theta != 0:
+            # Rotate the points to the expected coordinate system.
+            ctheta = np.cos(theta)
+            stheta = np.sin(theta)
+
+            mat = np.array([[ctheta, stheta],
+                            [-stheta, ctheta]])
+
+            # Calculate the points that could possibly be inside the volume
+            # Since the points are rotated, we cannot do anything directional
+            # without more logic
+            max_radius = np.abs(mat).dot(np.sqrt([a_squared, b_squared]))
+            idx, shapes = _getshapes_2d(center, max_radius, space.shape)
+
+            subgrid = [g[idi] for g, idi in zip(grid, shapes)]
+            offset_points = [vec * (xi - x0i)[..., None]
+                             for xi, vec, x0i in zip(subgrid, mat.T, [x0, y0])]
+            rotated = offset_points[0] + offset_points[1]
+            np.square(rotated, out=rotated)
+            max_distance = np.maximum(scales[0] * rotated[..., 0],
+                                      scales[1] * rotated[..., 1])
+        else:
+            # Calculate the points that could possibly be inside the volume
+            max_radius = np.sqrt([a_squared, b_squared])
+            idx, shapes = _getshapes_2d(center, max_radius, space.shape)
+
+            subgrid = [g[idi] for g, idi in zip(grid, shapes)]
+            squared_dist = [ai * (xi - x0i) ** 2
+                            for xi, ai, x0i in zip(subgrid,
+                                                   scales,
+                                                   [x0, y0])]
+
+            # Parentheses to get best order for broadcasting
+            max_distance = np.maximum(squared_dist[0], squared_dist[1])
+
+        # Find the points within the ellipse
+        inside = max_distance <= 1
+
+        # Add the ellipse intensity to those points
+        p[idx][inside] += intensity
+
+    return space.element(p)
+
+
+def cuboid_phantom(space, cuboids, min_pt=None, max_pt=None):
+    """Return a phantom given by cuboids.
+
+    Parameters
+    ----------
+    space : `DiscreteLp`
+        Space in which the phantom should be created, must be 2- or
+        3-dimensional. If ``space.shape`` is 1 in an axis, a corresponding
+        slice of the phantom is created (instead of squashing the whole
+        phantom into the slice).
+    cuboids : sequence of sequences
+        If ``space`` is 2-dimensional, each row should contain the entries ::
+
+            'value',
+            'axis_1', 'axis_2',
+            'center_x', 'center_y',
+            'rotation'
+
+        The provided cuboids need to be specified relative to the
+        reference rectangle ``[-1, -1] x [1, 1]``.
+        The angles are to be given in radians.
```

Comment:
How about 3D?

Reply:
3D is not currently supported, I'll update the rest to reflect this